pkg/workflows/wasm/host/module.go

```diff
@@ -429,19 +424,177 @@ func fetchFn(logger logger.Logger, modCfg *ModuleConfig) func(caller *wasmtime.C
 			return ErrnoFault
 		}
 
-		size := copyBuffer(caller, respBytes, respptr, int32(len(respBytes)))
-		if size == -1 {
+		if size := wasmWrite(caller, respBytes, respptr, int32(len(respBytes))); size == -1 {
 			return ErrnoFault
 		}
 
-		uint32Size := int32(4)
-		resplenBytes := make([]byte, uint32Size)
-		binary.LittleEndian.PutUint32(resplenBytes, uint32(len(respBytes)))
-		size = copyBuffer(caller, resplenBytes, resplenptr, uint32Size)
-		if size == -1 {
+		if size := wasmWriteUInt32(caller, resplenptr, uint32(len(respBytes))); size == -1 {
 			return ErrnoFault
 		}
 
 		return ErrnoSuccess
 	}
 }
+
+// createEmitFn injects dependencies and builds the emit function exposed by the WASM.  Errors in
+// Emit, if any, are returned in the Error Message of the response.
+func createEmitFn(
+	l logger.Logger,
+	e sdk.Emitter,
+	reader UnsafeReaderFunc,
+	writer UnsafeWriterFunc,
+	sizeWriter UnsafeFixedLengthWriterFunc,
+) func(caller *wasmtime.Caller, respptr, resplenptr, msgptr, msglen int32) int32 {
+	logErr := func(err error) {
+		l.Errorf("error emitting message: %s", err)
+	}
+
+	return func(caller *wasmtime.Caller, respptr, resplenptr, msgptr, msglen int32) int32 {
+		// writeErr marshals and writes an error response to wasm
+		writeErr := func(err error) int32 {
+			logErr(err)
+
+			resp := &wasmpb.EmitMessageResponse{
+				Error: &wasmpb.Error{
+					Message: err.Error(),
+				},
+			}
+
+			respBytes, perr := proto.Marshal(resp)
+			if perr != nil {
+				logErr(perr)
+				return ErrnoFault
+			}
+
+			if size := writer(caller, respBytes, respptr, int32(len(respBytes))); size == -1 {
+				logErr(errors.New("failed to write response"))
+				return ErrnoFault
+			}
+
+			if size := sizeWriter(caller, resplenptr, uint32(len(respBytes))); size == -1 {
+				logErr(errors.New("failed to write response length"))
+				return ErrnoFault
+			}
+
+			return ErrnoSuccess
+		}
+
+		b, err := reader(caller, msgptr, msglen)
+		if err != nil {
+			return writeErr(err)
+		}
+
+		msg := &wasmpb.EmitMessageRequest{}
+		err = proto.Unmarshal(b, msg)
+		if err != nil {
+			return writeErr(err)
+		}
+
+		vl, err := values.FromMapValueProto(msg.Labels)
+		if err != nil {
+			return writeErr(err)
+		}
+
+		// Handle the case of no labels before unwrapping.
+		if vl == nil {
+			vl = values.EmptyMap()
+		}
+
+		var labels map[string]any
+		if err := vl.UnwrapTo(&labels); err != nil {
+			return writeErr(err)
+		}
+
+		if err := e.Emit(msg.Message, labels); err != nil {
+			return writeErr(err)
+		}
+
+		return ErrnoSuccess
+	}
+}
+
+// beholderEmitter sends a message to the Beholder service via the custmsg package's labeler.  Each label is
+// expected to be a string.  Fails if any label is not a string.
+func beholderEmitter(msg string, labels map[string]any) error {
```

Comment:
IMO this consumer shows the impact of the current `Emit` interface on Labeler - the intent was for Labeler to be consumed at the highest level so that consumers can add labels at multiple depths in a call stack and pass it on.

Reply:
hm, not sure I follow, what's the suggested change/issue?

Reply:
It looks like you would have a consumer call `beholderEmitter`, which creates a new labeler with labels at that point of time, instead of creating a Labeler, adding labels through `With`, and then calling `Emit`. I could propose a specific change if you have an example of why `beholderEmitter` is needed.

Reply:
@patrickhuie19 IIUC correctly, you're saying we should preserve any existing labelers that are instantiated further up, is that correct?

This beholderEmitter would only be called if an emitter isn't passed in when calling NewModule. Where this is used I'd expect us to instantiate an emitter based off of an existing labeler.

Reply:
>IIUC correctly, you're saying we should preserve any existing labelers that are instantiated further up, is that correct?

Yes, and you wouldn't need a new func (`beholderEmitter`) if you had an instantiated Labeler and instead used its Emit func